File: 1/src/sum_multiples.erl
-module(sum_multiples).
-export([sum/2, solution/0]).

sum(Divisors, Under) ->
  Multiples = find_multiples(Divisors, Under - 1, []),
  lists:sum(Multiples).

find_multiples(_Divisors, 0, Multiples) ->
  Multiples;

find_multiples(Divisors, CurrentNumber, Multiples) ->
  case is_multiple_of(Divisors, CurrentNumber) of
    true  ->
      find_multiples(Divisors, CurrentNumber - 1, [CurrentNumber | Multiples]);
    false ->
      find_multiples(Divisors, CurrentNumber - 1, Multiples)
  end.

is_multiple_of(Divisors, Number) ->
  lists:any(fun(X) -> (Number rem X) =:= 0 end, Divisors).

solution() ->
  io:format(
    "~nThe sum of all the multiples of 3 or 5 below 1000: ~p~n",
    [sum([3,5], 1000)]
  ).

File: 1/sum_multiples.clj
(ns sum-multiples)

(defn is-multiple-of
  ([number] (fn [divisor] (= 0
                             (rem number divisor)))))

(defn is-multiple-of-some
  ([divisors number] (not= nil
                           (some (is-multiple-of number)
                                 divisors))))

(defn find-multiples [divisors limit]
  (loop [numbers (range 1 limit)
         multiples []]
    (if (and numbers)
      (recur (next numbers)
             (let [number (first numbers)]
               (if (is-multiple-of-some divisors number)
                 (conj multiples number)
                 multiples)))
      multiples)))

(defn sum-multiples [divisors number]
  (reduce + (find-multiples divisors number)))
